File: math/Quaternion.lua
--[[
-- ZHU Weixu
--   zhuweixu_harry@126.com
--
--   version 1.1 : some problems in create from 2 vecs are detected and marked
--   version 2   : create default from rotation, use createHardValue for hard set
--   version 3   : create getAxis, getAng
--   version 3.1 : create() creates a rotate((0,0,1),0) which is hardvalue(0,0,0,1) not a hardvalue(0,0,0,0)
--]]
local Vec3 = require("Vector3")

local Quaternion = {CLASS = "Quaternion", CLASSQUATERNION = true}
Quaternion.__index = Quaternion

function Quaternion:createFromHardValue(x,y,z,w)
	local instance = {}
	setmetatable(instance,self)
	self.__index = self

	if type(x) == "table" and x.CLASS == "Vector3" then
		instance.v = x
		instance.w = y
		return instance
	end
	if type(x) == "number" and 
	   type(y) == "number" and
	   type(z) == "number" and
	   type(w) == "number" then
		instance.v = Vec3:create(x,y,z)
		instance.w = w
		return instance
	end

	instance.v = Vec3:create(0,0,0)
	instance.w = 0
	return instance;
end

--[[  abandon this function
function Quaternion:createFrom4Vecs(_abc_o,_pqr_o,_abc,_pqr)
	-- give 4 vectors, rotation from the from the first two to last two
	-- some problem for specific situation
	-- 		only general case and
	-- 		only work for right angle and pqr_o = pqr
	local abc = _abc:nor()
	local pqr = _pqr:nor()
	local abc_o = _abc_o:nor()
	local pqr_o = _pqr_o:nor()

	local axis = (abc-abc_o) * (pqr - pqr_o)
	axis = axis:nor()

	local rot_o = abc_o - axis ^ abc * axis
	local rot_d = abc - axis ^ abc * axis
	rot_o = rot_o:nor()
	rot_d = rot_d:nor()
	local cos = rot_o ^ rot_d
	axis = rot_o * rot_d
	local th = math.acos(cos)

	local quater = self:create(axis,th)
	return quater
end
--]]

function Quaternion:create(x,y,z,th)	-- create from rotation
	local halfth
	local v
	if type(x) == "table" and x.CLASS == "Vector3" and x:len() ~= 0 then
		halfth = y / 2
		v = x / x:len()
		v = x:nor()
		return self:createFromHardValue(v * math.sin(halfth),math.cos(halfth))
	end
	if type(x) == "number" and
	   type(y) == "number" and
	   type(z) == "number" and
	   type(th) == "number" and
	   (x ~=0 or y ~= 0 or z ~= 0)then
		halfth = th / 2
		v = Vec3:create(x,y,z)
		v = v:nor()
		return self:createFromHardValue(v * math.sin(halfth),math.cos(halfth))
	end
	--return self:createFromHardValue(0,0,0,0)
	return self:create(0,0,1,0)
end

function Quaternion:getAxis()
	local halfth = math.acos(self.w)
	if halfth == 0 then
		return Vec3:create(0,0,1)
	end
	return (self.v / math.sin(halfth)):nor()
end

function Quaternion:getAng()	-- in rad
	return math.acos(self.w) * 2
end

function Quaternion.__add(a,b)
	return Quaternion:createFromHardValue(a.v + b.v,a.w + b.w)
end

function Quaternion.__unm(b)
	return Quaternion:createFromHardValue(-b.v,-b.w)
end

function Quaternion.__sub(a,b)
	local c = -b
	return a + c 
end

function Quaternion.__mul(a,b)
	local tv = a.v * b.v + a.w * b.v + a.v * b.w
	local tw = a.w * b.w - a.v ^ b.v
	return Quaternion:createFromHardValue(tv,tw)
end

function Quaternion:squlen()
	return self.v:squlen() + self.w * self.w
end

function Quaternion:len()
	return math.sqrt(self:squlen())
end

--inverse
function Quaternion:inv()
	if self:len() ~= 0 then
		return Quaternion:createFromHardValue(-self.v/self:len(),self.w/self:len())
	end
	return Quaternion:createFromHardValue(0,0,0,0)
end

function Quaternion:__tostring()
	local c = "(" .. "(" .. self.v.x .. "," .. self.v.y .. "," .. self.v.z .. ")"
	c = c .. "," .. self.w .. ")"
	return c
end

-- returns a vector, is a is not a vector, returns a whatever it is
function Quaternion:toRotate(a)
	if self.v:len() == 0 and self.w == 0 then
		return a
	end
	if type(a) == "table" and a.CLASS == "Vector3" then
		---[[
		local p = Quaternion:createFromHardValue(a,0)
		local res = self * p
		res = res * self:inv()
		local b = Vec3:create(res.v)
		--]]
		return b
	else
		print("In Quaternion:toRotate, para not a Vector3")
		return a
	end
end

return Quaternion
--function Quaternion

File: data/test6/error_calculator/CVector/Quaternion.cpp
/*----------------------------------------*/
/*	Weixu ZHU (Harry)
		zhuweixu_harry@126.com
	Version 2.0
		redesign Constructor
	Version 2.1
		fixed some bugs of the set or Quaternion(xxx)
*/
/*----------------------------------------*/

#include "Quaternion.h"
#include <stdio.h>	// for sprintf in toStr, and printf
#include <math.h>	// for sqrt

#define PI 3.1415926

/*-- Constructor Destructor-----------------------------*/
Quaternion::Quaternion()
	{ set(0,0,1,0); }
Quaternion::Quaternion(double _x, double _y, double _z, double _w)
	{ set(_x,_y,_z,_w); }
Quaternion::Quaternion(const Vector3& _l, double _w)
	{ set(_l,_w); }
Quaternion::Quaternion(const Quaternion& _x)
	{ set(_x); }

Quaternion::~Quaternion()
{
	//printf("i am a destroyer\n");
}

/*-- set ----------------------------*/
Quaternion& Quaternion::setHardValue(double _x, double _y, double _z, double _w)
{
	l.set(_x,_y,_z);
	w = _w;
	return *this;
}

Quaternion& Quaternion::setHardValue(const Vector3& _l, double _w)
{
	l.set(_l);
	w = _w;
	return *this;
}

Quaternion& Quaternion::set(double _x, double _y, double _z, double _w)
{
	return set(Vector3(_x,_y,_z),_w);
}

Quaternion& Quaternion::set(const Vector3& _l, double _w)
{
	double halfth = _w/2;
	l = _l.nor() * sin(halfth);
	w = cos(halfth);
	return *this;
}

Quaternion& Quaternion::set(const Quaternion& _x)
{
	l = _x.l;
	w = _x.w;
	return *this;
}

/*
Quaternion& Quaternion::setFrom4Vecs(const Vector3& _abc_o,const Vector3& _pqr_o,
							 const Vector3& _abc,  const Vector3& _pqr)
{
	//-- give 4 vectors, rotation from the from the first two to last two
	//-- some problem for specific situation
	//-- 		only general case and
	//-- 		only work for right angle and pqr_o = pqr
	Vector3 abc = _abc.nor();
	Vector3 pqr = _pqr.nor();
	Vector3 abc_o = _abc_o.nor();
	Vector3 pqr_o = _pqr_o.nor();

	Vector3 axis = (abc-abc_o) * (pqr - pqr_o);
	axis = axis.nor();

	Vector3 rot_o = abc_o - (axis ^ abc) * axis;
	Vector3 rot_d = abc - (axis ^ abc) * axis;
	rot_o = rot_o.nor();
	rot_d = rot_d.nor();
	double cos = rot_o ^ rot_d;
	axis = rot_o * rot_d;
	double th = acos(cos);

	this->setFromRotation(axis,th);
	return *this;
}
*/

/*-- get axis and ang----------------------------*/
Vector3 Quaternion::getAxis() const
{
	double halfth = acos(w);
	if (halfth == 0)
		return Vector3(0,0,1);
	Vector3 axis = (l / sin(halfth)).nor();
	return axis;
}
double Quaternion::getAng() const
{
	return acos(w) * 2;
}

/*-- operater ----------------------------*/
Quaternion Quaternion::operator^(double _x) const
{
	double th = acos(w) * 2 * _x;
	return Quaternion(getAxis(),th);
}

Quaternion& Quaternion::operator+=(const Quaternion& _x)
{
	l += _x.l;
	w += _x.w;
	return *this;
}

Quaternion& Quaternion::operator-=(const Quaternion& _x)
{
	l -= _x.l;
	w -= _x.w;
	return *this;
}

Quaternion  Quaternion::operator-() const
{
	Quaternion c;
	c -= *this;
	return c;
}

Quaternion  Quaternion::operator-(const Quaternion& _x) const
{
	Quaternion c(*this);
	c -= _x;
	return c;
}

Quaternion  Quaternion::operator+(const Quaternion& _x) const
{
	Quaternion c(*this);
	c += _x;
	return c;
}

Quaternion& Quaternion::operator*=(double _x)
{
	l *= _x;
	w *= _x;
	return *this;
}
Quaternion& Quaternion::operator/=(double _x)
{
	if (_x == 0)
	{
		printf("in Quaternion, /=, tried to divided by 0\n");
		this->set(0,0,1,0);
		return *this;
	}

	l /= _x;
	w /= _x;
	return *this;
}

Quaternion Quaternion::operator*(double _x) const
{
	Quaternion q;
	q.setHardValue(l*_x, w*_x);
	return q;
}

Quaternion Quaternion::operator/(double _x) const
{
	if (_x == 0)
	{
		printf("in Quaternion, /, tried to divided by 0\n");
		return Quaternion(0,0,1,0);
	}
	Quaternion q;
	q.setHardValue(l/_x, w/_x);
	return q;
}

Quaternion& Quaternion::operator*=(const Quaternion& _x)
{
	Quaternion c;
	c.setHardValue((this->l * _x.l) + (this->w * _x.l) + (this->l * _x.w),
				 (this->w * _x.w) - (this->l ^ _x.l));
	this->l = c.l;
	this->w = c.w;
	return *this;
}

Quaternion Quaternion::operator*(const Quaternion& _x) const
{
	Quaternion c;
	c.setHardValue((this->l * _x.l) + (this->w * _x.l) + (this->l * _x.w),
				 (this->w * _x.w) - (this->l ^ _x.l));
	return c;
}

double Quaternion::len() const
{
	return sqrt(l.x * l.x + l.y * l.y + l.z * l.z +
				w * w);
}

Quaternion Quaternion::inv() const
{
	double ll = len();
	if (ll == 0)
	{
		printf("in Quaternion, inv(), len == 0\n");
		return Quaternion(0,0,1,0);
	}

	Quaternion q;
	q.setHardValue(-l/ll,w/ll);
	return q;
}

/*---------- rotation -------------*/
Vector3 Quaternion::toRotate(const Vector3& _x) const
{
	Quaternion p;
	p.setHardValue(_x,0);
	Quaternion res = Quaternion(*this) * p;
	res = res * (this->inv());
	return res.l;
}

/*-- for print ----------------------------*/
char* Quaternion::toStr()
{
	sprintf(strForMe, "(%s, %lf)",l.toStr(),w);
	return strForMe;
}

